README.md
# Salesforce Marketing Cloud Cordova Plugin

Use this plugin to implement the Marketing Cloud MobilePush SDK for your [iOS](https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/) and [Android](http://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/) applications.

## Release Notes

Release notes for the plugin can be found [here](CHANGELOG.md)

## Installation

#### 1. Add plugin to your application via [npm](https://www.npmjs.com/package/cordova-plugin-marketingcloudsdk)

```shell
cordova plugin add cordova-plugin-marketingcloudsdk
```

**Required for iOS**

You must have [Cocoapods](https://cocoapods.org/) installed for the iOS platform integration.  Execute the following commands from terminal to install:
```shell
sudo gem install cocoapods
pod repo update
```

#### 2. Modify your application's `config.xml` to configure the plugin <a name="config"></a>

```xml
<!-- Required -->
<preference name="com.salesforce.marketingcloud.app_id" value="{Marketing Cloud application id}" />
<preference name="com.salesforce.marketingcloud.access_token" value="{Marketing Cloud access token}" />
<preference name="com.salesforce.marketingcloud.tenant_specific_endpoint" value="{URL retrieved from Marketing Cloud adminstration page}" />

<!-- Required - Android Only -->
<platform name="android">
  <preference name="com.salesforce.marketingcloud.notification_small_icon" value="ic_notification" />
</platform>

<!-- Optional -->
<preference name="com.salesforce.marketingcloud.analytics" value="{true|false}" />
<preference name="com.salesforce.marketingcloud.delay_registration_until_contact_key_is_set" value="{true|false}" />
```

#### 3. Provide FCM credentials

To enable push support for the Android platform you will need to include the google-services.json file.  

1. Download the file from your application's [Firebase console](https://console.firebase.google.com/) and place it in your project's root folder.  
2. Add following to Android element in your `config.xml`:

```xml
<platform name="android">
  <resource-file src="google-services.json" target="app/google-services.json" />
</platform>
```

#### 4. Enable iOS Push

Follow [these instructions](./ios_push.md) to enable push for iOS.

## API Reference <a name="reference"></a>


* [MCCordovaPlugin](#module_MCCordovaPlugin)
    * _static_
        * [.init(config, successCallback, [errorCallback])](#module_MCCordovaPlugin.init)
        * [.requestPushPermission(successCallback, [errorCallback])](#module_MCCordovaPlugin.requestPushPermission)
        * [.enableLocation(successCallback, [errorCallback])](#module_MCCordovaPlugin.enableLocation)
        * [.disableLocation(successCallback, [errorCallback])](#module_MCCordovaPlugin.disableLocation)
        * [.isLocationEnabled(successCallback, [errorCallback])](#module_MCCordovaPlugin.isLocationEnabled)
        * [.isPushEnabled(successCallback, [errorCallback])](#module_MCCordovaPlugin.isPushEnabled)
        * [.enablePush([successCallback], [errorCallback])](#module_MCCordovaPlugin.enablePush)
        * [.disablePush([successCallback], [errorCallback])](#module_MCCordovaPlugin.disablePush)
        * [.getSystemToken(successCallback, [errorCallback])](#module_MCCordovaPlugin.getSystemToken)
        * [.getAttributes(successCallback, [errorCallback])](#module_MCCordovaPlugin.getAttributes)
        * [.setAttribute(key, value, [successCallback], [errorCallback])](#module_MCCordovaPlugin.setAttribute)
        * [.clearAttribute(key, [successCallback], [errorCallback])](#module_MCCordovaPlugin.clearAttribute)
        * [.addTag(tag, [successCallback], [errorCallback])](#module_MCCordovaPlugin.addTag)
        * [.removeTag(tag, [successCallback], [errorCallback])](#module_MCCordovaPlugin.removeTag)
        * [.getTags(successCallback, [errorCallback])](#module_MCCordovaPlugin.getTags)
        * [.setContactKey(contactKey, [successCallback], [errorCallback])](#module_MCCordovaPlugin.setContactKey)
        * [.getContactKey(successCallback, [errorCallback])](#module_MCCordovaPlugin.getContactKey)
        * [.enableVerboseLogging([successCallback], [errorCallback])](#module_MCCordovaPlugin.enableVerboseLogging)
        * [.disableVerboseLogging([successCallback], [errorCallback])](#module_MCCordovaPlugin.disableVerboseLogging)
        * [.setOnNotificationOpenedListener(notificationOpenedListener)](#module_MCCordovaPlugin.setOnNotificationOpenedListener)
        * [.setOnUrlActionListener(urlActionListener)](#module_MCCordovaPlugin.setOnUrlActionListener)
        * [.logSdkState([successCallback], [errorCallback])](#module_MCCordovaPlugin.logSdkState)
        * [.track(eventName, attributesMap)](#module_MCCordovaPlugin.track)
    * _inner_
        * [~notificationOpenedCallback](#module_MCCordovaPlugin.notificationOpenedCallback) : <code>function</code>
        * [~urlActionCallback](#module_MCCordovaPlugin.urlActionCallback) : <code>function</code>


---

<a name="module_MCCordovaPlugin"></a>

## MCCordovaPlugin

<a name="module_MCCordovaPlugin.init"></a>

### MCCordovaPlugin.init(successCallback, [errorCallback])
Initialize the Marketing Cloud SDK programmatically.

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  
**See**
- [Android Docs](https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/8.0/com.salesforce.marketingcloud/-marketing-cloud-config/index.html)
- [iOS Docs](https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_configureWithDictionary:error:)

| Param | Type | Description |
| --- | --- | --- |
| config | <code>Object</code> | Key-value map of configuration values. |
| successCallback | <code>function</code> |  |
| [errorCallback] | <code>function</code> |  |

<a name="module_MCCordovaPlugin.requestPushPermission"></a>

### MCCordovaPlugin.requestPushPermission(successCallback, [errorCallback])
Request the user permission (iOS only).

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  

| Param | Type | Description |
| --- | --- | --- |
| successCallback | <code>function</code> |  |
| [errorCallback] | <code>function</code> |  |

<a name="module_MCCordovaPlugin.enableLocation"></a>

### MCCordovaPlugin.enableLocation(successCallback, [errorCallback])
Enable location.

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  

| Param | Type | Description |
| --- | --- | --- |
| successCallback | <code>function</code> |  |
| [errorCallback] | <code>function</code> |  |

<a name="module_MCCordovaPlugin.disableLocation"></a>

### MCCordovaPlugin.disableLocation(successCallback, [errorCallback])
Disable location.

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  

| Param | Type | Description |
| --- | --- | --- |
| successCallback | <code>function</code> |  |
| [errorCallback] | <code>function</code> |  |

<a name="module_MCCordovaPlugin.isLocationEnabled"></a>

### MCCordovaPlugin.isLocationEnabled(successCallback, [errorCallback])
The current state of the location flag in the native Marketing Cloud SDK.

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  

| Param | Type | Description |
| --- | --- | --- |
| successCallback | <code>function</code> |  |
| successCallback.enabled | <code>boolean</code> | Whether location is enabled. |
| [errorCallback] | <code>function</code> |  |

<a name="module_MCCordovaPlugin.isPushEnabled"></a>

### MCCordovaPlugin.isPushEnabled(successCallback, [errorCallback])
The current state of the pushEnabled flag in the native Marketing Cloud
SDK.

**Kind**: static method of [<code>MCCordovaPlugin</code>](#module_MCCordovaPlugin)  
**See**


CHANGELOG.md
Change Log
==========
### Version 7.5.0-c
* Adds init, geofencing and types files
### Version 7.5.0
* 7.4+ versions of the Android and iOS Marketing Cloud SDK are used in this version
* Adds Behavioral Triggers for InApp Message functionality, see `track()`

### Version 7.4.3
* 7.x versions of the Android and iOS Marketing Cloud SDK are used in this version
* Fix Insecure Maven URL for downloading of Marketing Cloud SDK in Android

### Version 7.4.2
* 7.x versions of the Android and iOS Marketing Cloud SDK are used in this version
* This plugin version requires Cordova Android 9.0.0 as minimum version to build , which makes minimum supported Android version as 5.1. 
* Google Play Services Plugin enable/disable support and version change can be modified in config.xml
* Added AndroidXEnabled preference as true , as Android Marketing Cloud SDK depends on these support libraries

### Version 7.4.1
* Updated to the 7.4.x versions of the Android and iOS Marketing Cloud SDK.
* Fix for removing un necessary Logcat statements on newIntent() flow in Android

### Version 7.4.0
* Updated to the 7.4.x version of the iOS Marketing Cloud SDK.

### Version 7.3.0
* Updated to the 7.3.x versions of the Android and iOS Marketing Cloud SDK.

### Version 7.2.0
* Updated to the 7.2.x versions of the Android and iOS Marketing Cloud SDK.

### Version 7.1.0
* Updated to the 7.1.x versions of the Android and iOS Marketing Cloud SDK.

### Version 7.0.0
* Updated to the 7.0.x versions of the Android and iOS Marketing Cloud SDK.

### Version 6.4.0
* Updated to the 6.4.x versions of the Android and iOS Marketing Cloud SDK.

### Version 6.3.1
* Added support for logging the SDK state to the native platform logging system (Android: Logcat, iOS: Console).
* Added documentation for setting up push for iOS.

### Version 6.3.0

* Updated to the 6.3.x versions of the Android and iOS Marketing Cloud SDK.
* The plugin now utilizes Cocoapods for the iOS SDK integration.
* Support In-App Messaging functionality of the MobilePush SDK.

### Version 6.2.1

* Updated to the 6.2.x versions of the Android and iOS Marketing Cloud SDK.
* Added support to delay registration until contact key has been set. Refer to the [config documentation](README.md#config) for more details.
* Require Tenant Specific Endpoint in the  `config.xml` file. If a Tenant Specific Endpoint is not specified an exception is thrown.

### Version 6.1.0

* Updated to the 6.1.x versions of the Android and iOS Marketing Cloud SDK.
* Added support for listening to notification opened events.  This will allow you to take action when a user clicks a notification, such as loading the url from an Open Direct or Cloud Page message.  

### Version 6.0.0

In this release we have greatly simplified the Cordova plugin for our MobilePush SDK.  For greatest success, if you are upgrading from a prior release, we suggest following the integration instructions as if this were a new plugin.

All functionality of this plugin remains the same as the preceding versions.

* The plugin version number will now match with the `major`.`minor` version of the SDK release.
* The plugin is now published to npm for easier installation
* The Javascript API has been updated to allow for optional parameters where applicable.  Refer to the [API Reference](README.md#reference) for more details.
* Moved SDK configuration values from command line arguments passed in while installing the SDK into the `config.xml` file.

### Version 1.1.0
_Released July 11, 2018_
> For Android: Depends on Android Cordova platform 6.4.0.

* Updated Android SDK to 5.6.x
* Updated iOS SDK to 5.2.x

### Version 1.0.3
_Released Apr 23, 2018_
> For iOS: Depends on the Marketing Cloud Mobile Push iOS SDK v5.1.x

www/MCCordovaPlugin.js
/**
 * @license
 * Copyright 2018 Salesforce, Inc
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

var exec = require('cordova/exec');
var argsCheck = require('cordova/argscheck');

var PLUGIN_NAME = 'MCCordovaPlugin';

var onNotificationOpened;
var onUrlAction;

function registerEvents() {
    var onEventsCallback = function(event) {
        switch (event.type) {
            case 'notificationOpened':
                if (onNotificationOpened !== undefined) {
                    onNotificationOpened(event);
                }
                break;
            case 'urlAction':
                if (onUrlAction !== undefined) {
                    onUrlAction(event);
                }
        }
    };

    _exec(onEventsCallback, null, 'registerEventsChannel');
}

document.addEventListener('deviceready', registerEvents);

function _exec(successCallback, errorCallback, methodName, args) {
    args = args || [];
    exec(successCallback, errorCallback, PLUGIN_NAME, methodName, args);
}

/**
 * @exports MCCordovaPlugin
 */
var MCCordovaPlugin = {

    /**
     * Initialize the Marketing Cloud SDK programmatically
     * @param {Object} config 
     * @param {function} successCallback 
     * @param {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/8.0/com.salesforce.marketingcloud/-marketing-cloud-config/index.html | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_configureWithDictionary:error: | iOS Docs}
     */
    init: function(config, successCallback, errorCallback = undefined) {
        argsCheck.checkArgs('ofF', `${PLUGIN_NAME}.init`, arguments);
        _exec(successCallback, errorCallback, 'init', [config]);
    },

    /**
     * Request the user permission (iOS only)
     * @param {function} successCallback 
     * @param {function} [errorCallback]
     */
    requestPushPermission: function(successCallback, errorCallback) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.requestPushPermission`, arguments);
        _exec(successCallback, errorCallback, 'requestPushPermission');
    },

    /**
     * Enable location
     * @param  {function} successCallback
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/7.0/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#isPushEnabled() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_locationEnabled | iOS Docs}
     */
    enableLocation: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.enableLocation`, arguments);
        _exec(successCallback, errorCallback, 'enableLocation');
    },

    /**
     * Disable location
     * @param  {function} successCallback
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/7.0/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#isPushEnabled() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_locationEnabled | iOS Docs}
     */
    disableLocation: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.disableLocation`, arguments);
        _exec(successCallback, errorCallback, 'disableLocation');
    },

    /**
     * The current state of the location flag in the native Marketing Cloud
     * SDK.
     * @param  {function(enabled)} successCallback
     * @param  {boolean} successCallback.enabled - Whether location is enabled.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/7.0/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#isPushEnabled() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_locationEnabled | iOS Docs}
     */
    isLocationEnabled: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.isLocationEnabled`, arguments);
        _exec(successCallback, errorCallback, 'isLocationEnabled');
    },

    /**
     * The current state of the pushEnabled flag in the native Marketing Cloud
     * SDK.
     * @param  {function(enabled)} successCallback
     * @param  {boolean} successCallback.enabled - Whether push is enabled.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#isPushEnabled() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_pushEnabled | iOS Docs}
     */
    isPushEnabled: function(successCallback, errorCallback = undefined) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.isPushEnabled`, arguments);
        _exec(successCallback, errorCallback, 'isPushEnabled');
    },
    /**
     * Enables push messaging in the native Marketing Cloud SDK.
     * @param  {function} [successCallback]
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#enablePush() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setPushEnabled: | iOS Docs}
     */
    enablePush: function(successCallback = undefined, errorCallback = undefined) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.enablePush`, arguments);
        _exec(successCallback, errorCallback, 'enablePush');
    },
    /**
     * Disables push messaging in the native Marketing Cloud SDK.
     * @param  {function} [successCallback]
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#disablePush() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setPushEnabled: | iOS Docs}
     */
    disablePush: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.disablePush`, arguments);
        _exec(successCallback, errorCallback, 'disablePush');
    },
    /**
     * Returns the token used by the Marketing Cloud to send push messages to
     * the device.
     * @param  {function(token)} successCallback
     * @param  {string} successCallback.token - The token used for push
     *     messaging.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/messages/push/PushMessageManager.html#getPushToken() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_deviceToken | iOS Docs}
     */
    getSystemToken: function(successCallback, errorCallback) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.getSystemToken`, arguments);
        _exec(successCallback, errorCallback, 'getSystemToken');
    },
    /**
     * Returns the maps of attributes set in the registration.
     * @param  {function(attributes)} successCallback
     * @param  {Object.<string, string>} successCallback.attributes - The
     *     key/value map of attributes set in the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.html#getAttributes() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_attributes | iOS Docs}
     */
    getAttributes: function(successCallback, errorCallback) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.getAttributes`, arguments);
        _exec(successCallback, errorCallback, 'getAttributes');
    },
    /**
     * Sets the value of an attribute in the registration.
     * @param  {string} key - The name of the attribute to be set in the
     *     registration.
     * @param  {string} value - The value of the `key` attribute to be set in
     *     the registration.
     * @param  {function(saved)} [successCallback]
     * @param  {boolean} successCallback.saved - Whether the attribute value was
     *     set in the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.Editor.html#setAttribute(java.lang.String,%20java.lang.String) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setAttributeNamed:value: | iOS Docs}
     */
    setAttribute: function(key, value, successCallback, errorCallback) {
        argsCheck.checkArgs('ssFF', `${PLUGIN_NAME}.setAttribute`, arguments);
        _exec(successCallback, errorCallback, 'setAttribute', [key, value]);
    },
    /**
     * Clears the value of an attribute in the registration.
     * @param  {string} key - The name of the attribute whose value should be
     *     cleared from the registration.
     * @param  {function(saved)} [successCallback]
     * @param  {boolean} successCallback.saved - Whether the value of the `key`
     *     attribute was cleared from the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.Editor.html#clearAttribute(java.lang.String) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_clearAttributeNamed: | iOS Docs}
     */
    clearAttribute: function(key, successCallback, errorCallback) {
        argsCheck.checkArgs('sFF', `${PLUGIN_NAME}.clearAttribute`, arguments);
        _exec(successCallback, errorCallback, 'clearAttribute', [key]);
    },
    /**
     * @param  {string} tag - The tag to be added to the list of tags in the
     *     registration.
     * @param  {function(saved)} [successCallback]
     * @param  {boolean} successCallback.saved - Whether the value passed in for
     *     `tag` was saved in the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.Editor.html#addTag(java.lang.String) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_addTag: | iOS Docs}
     */
    addTag: function(tag, successCallback, errorCallback) {
        argsCheck.checkArgs('sFF', `${PLUGIN_NAME}.addTag`, arguments);
        _exec(successCallback, errorCallback, 'addTag', [tag]);
    },
    /**
     * @param  {string} tag - The tag to be removed from the list of tags in the
     *     registration.
     * @param  {function(saved)} [successCallback]
     * @param  {boolean} successCallback.saved - Whether the value passed in for
     *     `tag` was cleared from the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.Editor.html#removeTag(java.lang.String) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_removeTag: | iOS Docs}
     */
    removeTag: function(tag, successCallback, errorCallback) {
        argsCheck.checkArgs('sFF', `${PLUGIN_NAME}.removeTag`, arguments);
        _exec(successCallback, errorCallback, 'removeTag', [tag]);
    },
    /**
     * Returns the tags currently set on the device.
     * @param  {function(tags)} successCallback
     * @param  {string[]} successCallback.tags - The array of tags currently set
     *     in the native SDK.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.html#getTags() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_tags | iOS Docs}
     */
    getTags: function(successCallback, errorCallback) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.getTags`, arguments);
        _exec(successCallback, errorCallback, 'getTags');
    },
    /**
     * Sets the contact key for the device's user.
     * @param  {string} contactKey - The value to be set as the contact key of
     *     the device's user.
     * @param  {function(saved)} [successCallback]
     * @param  {boolean} successCallback.saved - Whether the value passed in for
     *     `contactKey` was saved in the registration.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.Editor.html#setContactKey(java.lang.String) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setContactKey: | iOS Docs}
     */
    setContactKey: function(contactKey, successCallback, errorCallback) {
        argsCheck.checkArgs('sFF', `${PLUGIN_NAME}.setContactKey`, arguments);
        _exec(successCallback, errorCallback, 'setContactKey', [contactKey]);
    },
    /**
     * Returns the contact key currently set on the device.
     * @param  {function(contactKey)} successCallback
     * @param  {string} successCallback.contactKey - The current contact key.
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/registration/RegistrationManager.html#getContactKey() | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_contactKey | iOS Docs}
     */
    getContactKey: function(successCallback, errorCallback) {
        argsCheck.checkArgs('fF', `${PLUGIN_NAME}.getContactKey`, arguments);
        _exec(successCallback, errorCallback, 'getContactKey');
    },
    /**
     * Enables verbose logging within the native Marketing Cloud SDK.
     * @param  {function} [successCallback]
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/MarketingCloudSdk.html#setLogLevel(int) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setDebugLoggingEnabled: | iOS Docs}
     */
    enableVerboseLogging: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.enableVerboseLogging`, arguments);
        _exec(successCallback, errorCallback, 'enableVerboseLogging');
    },
    /**
     * Disables verbose logging within the native Marketing Cloud SDK.
     * @param  {function} [successCallback]
     * @param  {function} [errorCallback]
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/6.4/reference/com/salesforce/marketingcloud/MarketingCloudSdk.html#setLogLevel(int) | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/MarketingCloudSDK.html#//api/name/sfmc_setDebugLoggingEnabled: | iOS Docs}
     */
    disableVerboseLogging: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.disableVerboseLogging`, arguments);
        _exec(successCallback, errorCallback, 'disableVerboseLogging');
    },
    /**
     * @param {function(event)} notificationOpenedListener
     * @param {MCCordovaPlugin~notificationOpenedCallback}
     *     notificationOpenedListener.event
     * @since 6.1.0
     */
    setOnNotificationOpenedListener: function(notificationOpenedListener) {
        argsCheck.checkArgs('f', `${PLUGIN_NAME}.setOnNotificationOpenedListener`, arguments);
        onNotificationOpened = notificationOpenedListener;
        _exec(undefined, undefined, 'subscribe', ['notificationOpened']);
    },

    /**
     * @callback module:MCCordovaPlugin~notificationOpenedCallback
     * @param {number} timeStamp - Time since epoch when the push message was
     *     opened.
     * @param {Object} values - The values of the notification message.
     * @param {string} values.alert - The alert text of the notification
     *     message.
     * @param {string} [values.title] - The title text of the notification
     *     message.
     * @param {string} [values.url] - The url associated with the notification
     *     message. This can be either a cloud-page url or an open-direct url.
     * @param {string} values.type - Indicates the type of notification message.
     *     Possible values: 'cloudPage', 'openDirect' or 'other'
     */

    /**
     * @param {function(event)} urlActionListener
     * @param {MCCordovaPlugin~urlActionCallback} urlActionListener.event
     * @since 6.3.0
     */
    setOnUrlActionListener: function(urlActionListener) {
        argsCheck.checkArgs('f', `${PLUGIN_NAME}.setOnUrlActionListener`, arguments);
        onUrlAction = urlActionListener;
        _exec(undefined, undefined, 'subscribe', ['urlAction']);
    },

    /**
     * @callback module:MCCordovaPlugin~urlActionCallback
     * @param {string} url - The url associated with the action taken by the user.
     */

    /**
     * Instructs the native SDK to log the SDK state to the native logging system (Logcat for
     * Android and Xcode/Console.app for iOS).  This content can help diagnose most issues within
     * the SDK and will be requested by the Marketing Cloud support team.
     *
     * @param  {function} [successCallback]
     * @param  {function} [errorCallback]
     * @since 6.3.1
     */
    logSdkState: function(successCallback, errorCallback) {
        argsCheck.checkArgs('FF', `${PLUGIN_NAME}.logSdkState`, arguments);
        _exec(successCallback, errorCallback, 'logSdkState');
    },

    /**
     * Method to track events, which could result in actions such as an InApp Message being
     * displayed.
     * @param  {string} eventName - The name of the event to be tracked.
     * @param  {string} attributesMap - key-value pairs of attributes associated with the Event.
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-Android/javadocs/MarketingCloudSdk/7.4/com.salesforce.marketingcloud.events/-event-manager/custom-event.html | Android Docs}
     * @see  {@link https://salesforce-marketingcloud.github.io/MarketingCloudSDK-iOS/appledoc/Classes/SFMCEvent.html#/c:objc(cs)SFMCEvent(cm)customEventWithName:withAttributes: | iOS Docs}
     */
    track: function(eventName, attributesMap) {
        argsCheck.checkArgs('soFF', `${PLUGIN_NAME}.track`, arguments);
        _exec(undefined, undefined, 'track', [eventName, attributesMap]);
    }

};

module.exports = MCCordovaPlugin;
